src/game.cpp

```diff
@@ -1533,16 +1533,32 @@ bool game::do_turn()
 
             mon_info_update();
 
-            // If player is performing a task and a monster is dangerously close, warn them
-            // regardless of previous safemode warnings
+            // If player is performing a task, a monster is dangerously close,
+            // and monster can reach to the player or it has some sort of a ranged attack,
+            // warn them regardless of previous safemode warnings
             if( u.activity && !u.has_activity( activity_id( "ACT_AIM" ) ) &&
                 u.activity.moves_left > 0 &&
                 !u.activity.is_distraction_ignored( distraction_type::hostile_spotted_near ) ) {
                 Creature *hostile_critter = is_hostile_very_close();
+
+                bool ranged_attack_monster = false;
                 if( hostile_critter != nullptr ) {
-                    cancel_activity_or_ignore_query( distraction_type::hostile_spotted_near,
-                                                     string_format( _( "The %s is dangerously close!" ),
-                                                             hostile_critter->get_name() ) );
+                    for( const std::pair<std::string, mtype_special_attack> &attack :
+                         hostile_critter->as_monster()->type->special_attacks ) {
+                        if( attack.second->id == "gun" ) {
+                            ranged_attack_monster = true;
+                            break;
+                        }
+                    }
+
+                    const std::vector<tripoint> path = get_map().route( hostile_critter->pos(), u.pos(),
```

Comment:
Couple of things here:
1. don't use the monster settings here. Create your own `pathfinding_settings` with max distance and length that correspond to the "danger range" (pathfinding is expensive!), and err on the side of the caution: assume monsters can open doors and walk over the traps.
2. probably swap `hostile_critter->pos()`, `u.pos()`, so path is calculated from `u` to the `critter`. This could be a bit more cache friendly for the repeated invocation
3. create a static empty set of tripoints and pass it as a third param

Reply:
Ok, done all that.